xapian-letor/ranker/listmle_ranker.cc

```diff
@@ -0,0 +1,211 @@
+/** @file listmle_ranker.cc
+ *  @brief Implementation of ListMLERanker
+ */
+/* Copyright (C) 2014 Hanxiao Sun
+ * Copyright (C) 2016 Ayush Tomar
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License as
+ * published by the Free Software Foundation; either version 2 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301
+ * USA
+ */
+
+/*
+ListMLE is adapted from the paper:
+Fen Xia, Tie-Yan Liu, Jue Wang, Wensheng Zhang, Hang Li
+et al. "Listwise Approach to Learning to Rank - Theory and Algorithm."
+*/
+#include <config.h>
+
+#include "xapian-letor/ranker.h"
+
+#include "debuglog.h"
+#include "serialise-double.h"
+
+#include <xapian.h>
+#include <xapian/intrusive_ptr.h>
+#include <xapian/types.h>
+#include <xapian/visibility.h>
+
+#include <algorithm>
+#include <cmath>
+#include <stdlib.h>
+
+using namespace std;
+using namespace Xapian;
+
+ListMLERanker::~ListMLERanker()
+{
+    LOGCALL_DTOR(API, "ListMLERanker");
+}
+
+static bool
+label_comparer(const FeatureVector& firstfv,
+	       const FeatureVector& secondfv)
+{
+    return firstfv.get_label() > secondfv.get_label();
+}
+
+static double
+calculate_inner_product(const vector<double>& parameters,
+			const vector<double>& feature_sets)
+{
+    double inner_product = 0.0;
+    for (size_t i = 0; i < parameters.size(); ++i) {
+	inner_product += parameters[i] * feature_sets[i];
+    }
+    return inner_product;
+}
+
+static vector<double>
+calculate_gradient(const vector<FeatureVector>& sorted_feature_vectors,
+		   const vector<double>& new_parameters)
+{
+    //need to be optimized,how to get the length of the features
+    vector<double> gradient(sorted_feature_vectors[0].get_fcount(), 0);
+
+    size_t list_length = sorted_feature_vectors.size();
+
+    vector<double> exponents;
+    double expsum = 0.0;
+
+    for (size_t i = 0; i < list_length; ++i) {
+	double temp = exp(calculate_inner_product(
+			      new_parameters,
+			      sorted_feature_vectors[i].get_fvals()));
```

Comment:
This kind of line wrapping doesn't come up very often, so I'm not 100% sure if we even have a consistent style, but I'd put the indented lines at lot less indented, possibly only 4 characters in (ie one indent) from the "parent" line. (You can then put `new_parameters` and the `sorted_feature_vectors…` expression on the same line.)

As it stands, scanning down the file my brain comes to an abrupt stop because of the enormous indentation gap, rather than reading it as a single (albeit nested) function call.